devito/profiler.py

```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+
+    def _get_for_flops(self, name, loop, loads):
+        loop_flops = 0
+        loop_oi_f = 0
+
+        if isinstance(loop.body, Assign):
+            loop_flops = self._get_assign_flops(loop.body, loads)
+            loop_oi_f = loop_flops
+        elif isinstance(loop.body, Block):
+            loop_oi_f, loop_flops = self._get_block_oi_and_flops(name, loop.body, loads)
+        elif isinstance(loop.body, For):
+            loop_oi_f = self._get_for_flops(name, loop.body, loads)
+
+        if loop_flops == 0:
+            return loop_oi_f
+
+        flops_calc_stmt = Statement("%s%s += %f" % (self.loop_temp_prefix, name, loop_flops)) if self.openmp else\
+            Statement("%s->%s += %f" % (self.f_name, name, loop_flops))
+        loop.body = Block([flops_calc_stmt, loop.body])
+
+        return loop_oi_f
+
+    def _get_block_oi_and_flops(self, name, block, loads):
+        block_flops = 0
+        block_oi = 0
+
+        for elem in block.contents:
+            if isinstance(elem, Assign):
+                a_flops = self._get_assign_flops(elem, loads)
+                block_flops += a_flops
+                block_oi += a_flops
+            elif isinstance(elem, Block):
+                nblock_oi, nblock_flops = self._get_block_oi_and_flops(name, elem, loads)
+                block_oi += nblock_oi
+                block_flops += nblock_flops
+            elif isinstance(elem, For):
+                block_oi += self._get_for_flops(name, elem, loads)
+
+        return block_oi, block_flops
+
+    def _get_assign_flops(self, assign, loads):
+        flops = 0
+        cur_load = ""
+        loads["stores"] += 1
+
+        idx = 0
+        brackets = 0
+        # removing casting statements and function calls to floor that can confuse the parser
+        string = (assign.lvalue + " " + assign.rvalue).replace("float", '').replace("int", '').replace("floor", '')
+
+        while idx < len(string):
+            char = string[idx]
+            if len(cur_load) == 0:
+                if char == '[':
+                    brackets += 1
+                elif char == ']':
+                    brackets -= 1
+                elif char in "+-*/" and string[idx - 1] is not 'e' and not string[idx + 1].isdigit() and brackets == 0:
+                    flops += 1
+                elif char.isalpha() and not string[idx - 1].isdigit() and char not in "it":
+                    cur_load += char
+                idx += 1
+            else:
+                if char is '[':
+                    loads[cur_load] = True
+                    cur_load = ""
+                    brackets += 1
+                    idx += 1
+                elif char is ' ' and brackets == 0 and len(cur_load) > 0:
+                    loads[cur_load] = True
+                    cur_load = ""
+                    idx += 1
+                else:
+                    cur_load += char
+                    idx += 1
```

Comment:
The method is still hacky, though

The right way of doing this, I think, is to have a visitor that traverses cgen expressions, not analysing strings textually